File: bindings/typescript/scripts/stage-native.mjs
#!/usr/bin/env node
/**
 * Copy the napi cdylib into this package as aidb.node / aidb.<os>-<arch>.node.
 * Used by npm pack, prepare, and CI. Users run `npm i aidb` — they do not copy a dylib.
 */
import { spawnSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const here = path.dirname(fileURLToPath(import.meta.url));
const pkg = path.resolve(here, "..");
const repo = path.resolve(pkg, "../..");

function platformTag() {
  const os =
    process.platform === "darwin"
      ? "darwin"
      : process.platform === "linux"
        ? "linux"
        : process.platform === "win32"
          ? "win32"
          : null;
  const arch =
    process.arch === "arm64" ? "arm64" : process.arch === "x64" ? "x64" : null;
  if (!os || !arch) {
    throw new Error(
      `aidb has no prebuilt napi addon for ${process.platform}-${process.arch}. Install from the repo with a Rust toolchain.`
    );
  }
  return `${os}-${arch}`;
}

function cargoNames() {
  if (process.platform === "darwin") {
    return ["libaidb_node.dylib", "aidb.node"];
  }
  if (process.platform === "win32") {
    return ["aidb_node.dll", "aidb.node"];
  }
  return ["libaidb_node.so", "aidb.node"];
}

function targetRoots() {
  const roots = [];
  if (process.env.CARGO_TARGET_DIR) {
    roots.push(process.env.CARGO_TARGET_DIR);
  }
  roots.push(path.join(repo, "target"));
  return roots;
}

function findArtifact(profile) {
  const names = cargoNames();
  for (const root of targetRoots()) {
    for (const name of names) {
      const candidate = path.join(root, profile, name);
      if (fs.existsSync(candidate)) {
        return candidate;
      }
    }
  }
  return null;
}

function resolveProfile(wantRelease) {
  if (wantRelease) {
    return "release";
  }
  const env = process.env.AIDB_NATIVE_PROFILE;
  if (env === "debug" || env === "release") {
    return env;
  }
  if (findArtifact("release")) {
    return "release";
  }
  if (findArtifact("debug")) {
    return "debug";
  }
  return "release";
}

function cargoBuild(profile) {
  const args = ["build", "-p", "aidb-node"];
  if (profile === "release") {
    args.push("--release");
  }
  const result = spawnSync("cargo", args, {
    cwd: repo,
    stdio: "inherit",
  });
  if (result.status !== 0) {
    process.exit(result.status ?? 1);
  }
}

const wantRelease = process.argv.includes("--release");
const profile = resolveProfile(wantRelease);
let source = findArtifact(profile);
if (!source) {
  cargoBuild(profile);
  source = findArtifact(profile);
}
if (!source) {
  throw new Error("aidb-node build did not produce a cdylib");
}

const tag = platformTag();
const tagged = path.join(pkg, `aidb.${tag}.node`);
fs.copyFileSync(source, tagged);
console.log(`staged ${path.relative(repo, source)} -> ${path.relative(repo, tagged)}`);

File: bindings/python/scripts/stage_native.py
#!/usr/bin/env python3
"""Copy the PyO3 cdylib into the aidb package.

Used by pip/CI. Users run `pip install aidb` — they do not copy a dylib.
"""

from __future__ import annotations

import os
import shutil
import subprocess
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
PKG = HERE.parent
REPO = PKG.parent.parent
AIDB_PKG = PKG / "aidb"


def cargo_names() -> list[str]:
    if sys.platform == "darwin":
        return ["libaidb_native.dylib", "aidb_native.so", "aidb_native.abi3.so"]
    if os.name == "nt":
        return ["aidb_native.pyd", "aidb_native.dll", "libaidb_native.dll"]
    return ["libaidb_native.so", "aidb_native.so", "aidb_native.abi3.so"]


def dest_name() -> str:
    if os.name == "nt":
        return "aidb_native.pyd"
    return "aidb_native.abi3.so"


def target_roots() -> list[Path]:
    roots: list[Path] = []
    if os.environ.get("CARGO_TARGET_DIR"):
        roots.append(Path(os.environ["CARGO_TARGET_DIR"]))
    roots.append(REPO / "target")
    return roots


def find_artifact(profile: str) -> Path | None:
    names = cargo_names()
    for root in target_roots():
        folder = root / profile
        for name in names:
            candidate = folder / name
            if candidate.exists():
                return candidate
    return None


def resolve_profile(want_release: bool) -> str:
    if want_release:
        return "release"
    env = os.environ.get("AIDB_NATIVE_PROFILE")
    if env in {"debug", "release"}:
        return env
    if find_artifact("release"):
        return "release"
    if find_artifact("debug"):
        return "debug"
    return "release"


def cargo_build(profile: str) -> None:
    args = ["cargo", "build", "-p", "aidb-python"]
    if profile == "release":
        args.append("--release")
    subprocess.check_call(args, cwd=REPO)


def stage(want_release: bool = False) -> Path:
    profile = resolve_profile(want_release)
    source = find_artifact(profile)
    if source is None:
        cargo_build(profile)
        source = find_artifact(profile)
    if source is None:
        raise FileNotFoundError("aidb-python build did not produce a cdylib")
    AIDB_PKG.mkdir(parents=True, exist_ok=True)
    dest = AIDB_PKG / dest_name()
    shutil.copy2(source, dest)
    print(f"staged {source} -> {dest}")
    return dest


def main() -> None:
    want_release = "--release" in sys.argv
    stage(want_release=want_release)


if __name__ == "__main__":
    main()
